dune/iga/nurbstrimfunctionality.hh

```diff
@@ -0,0 +1,303 @@
+    auto d2 = p4 - p3;
+
+    auto cos_angle = Dune::dot(d1, d2) / (d1.two_norm() * d2.two_norm());
+    auto angle     = std::acos(std::min(std::max(cos_angle, -1.0), 1.0));
+
+    // Check if the angle is small enough to consider the curves on the same line
+    return ((angle < tolerance) || (std::fabs(angle - std::numbers::pi) < tolerance));
+  }
+
+  struct IntersectionResult {
+    enum ResultFlag { Found, NotFound };
+
+    double localResult{};
+    Point globalResult;
+    ResultFlag flag = NotFound;
+  };
+
+  IntersectionResult newtonRaphson(NurbsCurveHandler& curve, Point& intersectionPoint) {
+    const double tolerance2 = 1e-7;
+    const double tolerance  = 1e-4;
+
+    // There are some easy cases where the inital Guess is at the back or at the front of the domain, check these cases
+    // first
+    if ((curve(curve.domain().front()) - intersectionPoint).two_norm() < tolerance2)
+      return {curve.domain().front(), curve(curve.domain().front()), IntersectionResult::ResultFlag::Found};
+    if ((curve(curve.domain().back()) - intersectionPoint).two_norm() < tolerance2)
+      return {curve.domain().back(), curve(curve.domain().back()), IntersectionResult::ResultFlag::Found};
+
+    // At first, we assume that the intersectionPoint is in some definition exact and the corresponding u has to be
+    // found
+    Dune::FieldVector<double, 1> u{curve.domainMidPoint()};
+    Dune::FieldVector<double, 1> du;
+
+    // Set up the algorithm
+    Point dGlobal;
+    int max_iterations = 250;
+    int i              = 0;
+    do {
+      dGlobal = curve(u[0]) - intersectionPoint;
+
+      using MatrixHelper = typename Dune::MultiLinearGeometryTraits<double>::MatrixHelper;
+      MatrixHelper::template xTRightInvA<1, 2>(curve.jacobianTransposed(u[0]), dGlobal, du);
+
+      // Update the guess
+      u -= du;
+
+      ++i;
+      if (i > max_iterations) return {u[0], intersectionPoint, IntersectionResult::ResultFlag::NotFound};
+
+      // Clamp result, this might be already an indicator that there is no solution
+      if (Dune::FloatCmp::gt(u[0], curve.patchData->knotSpans[0].back())) u[0] = curve.patchData->knotSpans[0].back();
+      if (Dune::FloatCmp::lt(u[0], curve.patchData->knotSpans[0].front())) u[0] = curve.patchData->knotSpans[0].front();
+
+    } while (dGlobal.two_norm() > tolerance);
+
+    double localResult = u[0];
+    auto globalResult  = curve(localResult);
+
+    return {localResult, globalResult, IntersectionResult::ResultFlag::Found};
+  }
+
+  auto approximatelySamePointNOTExact = [](Point a, Point b) {
+    double tol               = 1e-2;
+    auto approximatelyEqual2 = [tol](auto a, auto b) { return fabs(a - b) < tol; };
+
+    return (approximatelyEqual2(a[0], b[0]) && approximatelyEqual2(a[1], b[1]));
+  };
+
+  auto getEdgeOrientation = [](int edge) -> int {
+    if (edge == 0 || edge == 2)
+      return 0;
+    else
+      return 1;
+  };
+  auto iPlusX = [](int i, int x) -> int { return (i + x) % 4; };
```

Comment:
These lambdas can be a usual functions.. Furthermore, ìPlusX` is not very descriptive